Use `referrerpolicy` attribute when navigating for `a`, `area`, `iframe` elements

estark37 opened the pull request:
Instead of initiating navigations with URLs, `a`, `area`, and `iframe` now create requests and set the referrer policy on these requests from the element's `referrerpolicy` attribute. These requests are passed to the navigation algorithm instead of just a URL.

estark37 commented:
One more for you to take a look at, @domenic. (I think this should now cover all the elements with `referrerpolicy` attributes.)

domenic commented on source:
```diff
@@ -22051,10 +22051,18 @@ interface <dfn>HTMLHyperlinkElementUtils</dfn> {
+   <li><p>Let <var>resource</var> be a new <span data-x="concept-request">request</span> whose <span
+   data-x="concept-request-url">url</span> is <var>URL</var>.</p></li>
+
+   <li><p>If <var>subject</var> is an <code>a</code> or <code>area</code> element, set
```
I think this should even apply for `<link>` elements, although it's pretty theoretical. The idea would be if a browser used e.g. `<link rel="next" href="..." referrerpolicy="..">` to display a "Go to next page" navigation UI, this "follows a hyperlink" algorithm would apply. In that case I think referrerpolicy should still work.

This would mean you'd need to change `<code data-x="attr-hyperlink-referrerpolicy">referrerpolicy</code>` to `<code data-x="">referrerpolicy</code>`, which is a bit sad, but it's something we do elsewhere for attributes shared between multiple elements.

domenic commented:
Looks great, with only one nit.

While you're here, would you mind reviewing https://github.com/whatwg/html/pull/1559 for me?

domenic commented:
Oh, and it would be good to add references to https://github.com/w3c/webappsec-referrer-policy/issues/40 and https://github.com/whatwg/html/issues/1531 to the commit messages.

estark37 commented:
Thanks @domenic. I updated "follows a hyperlink" to work for `link` elements, squashed, and referenced https://github.com/w3c/webappsec-referrer-policy/issues/40 and https://github.com/whatwg/html/issues/1531 in the squashed commit message. Will take a look at #1559 now.

domenic commented on source:
```diff
@@ -22051,10 +22051,18 @@ interface <dfn>HTMLHyperlinkElementUtils</dfn> {
+   <li><p>Let <var>resource</var> be a new <span data-x="concept-request">request</span> whose <span
+   data-x="concept-request-url">url</span> is <var>URL</var>.</p></li>
+
+   <li><p>If <var>subject</var> is an <code>a</code>, <code>area</code>, or <code>link</code>
```
You can just remove the "If" predicate entirely now and merge this bullet point into the previous one, I think.

estark37 replied:
Ah, good point, done.

domenic commented on source:
```diff
@@ -28132,8 +28137,14 @@ interface <dfn>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>resource</var> be a new <span data-x="concept-request">request</span> whose <span
+   data-x="concept-request-url">url</span> is <var>url</var> and whose <span
+   data-x="concept-request-referrer-policy">referrer policy</span> is the current state of the
+   element's <code data-x="attr-iframe-referrerpolicy">referrerpolicy</code> content
```
s/the element/_subject_

estark37 replied:
Done

estark37 replied:
Wait I think I changed the wrong one, sorry!

estark37 replied:
Ok, no, I think this is right? Following a hyperlink uses _subject_, but the otherwise steps for iframe use "the element" (consistent with the rest of the algorithm).

domenic replied:
Ah, yep, I commented on the wrong line, but you changed the right one!
